baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/Slf4jLevelCheck.java

```diff
@@ -0,0 +1,183 @@
+import com.sun.source.tree.IfTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.ParenthesizedTree;
+import com.sun.source.util.SimpleTreeVisitor;
+import com.sun.source.util.TreeScanner;
+import java.util.Locale;
+import java.util.Optional;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "Slf4jLevelCheck",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = SeverityLevel.ERROR,
+        summary = "Slf4j logger.is[Level]Enabled level must match the most severe log statement")
+public final class Slf4jLevelCheck extends BugChecker implements IfTreeMatcher {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Matcher<ExpressionTree> LEVEL_CHECK_METHOD = MethodMatchers.instanceMethod()
+            .onDescendantOf("org.slf4j.Logger")
+            .namedAnyOf("isTraceEnabled", "isDebugEnabled", "isInfoEnabled", "isWarnEnabled", "isErrorEnabled");
+
+    private static final Matcher<ExpressionTree> LOG_METHOD = MethodMatchers.instanceMethod()
+            .onDescendantOf("org.slf4j.Logger")
+            .namedAnyOf("trace", "debug", "info", "warn", "error");
+
+    @Override
+    public Description matchIf(IfTree tree, VisitorState state) {
+        // n.b. This check does not validate that the level check and logging occur on the same logger instance.
+        // It's possible to have multiple loggers in the same class used for different purposes, however we recommend
+        // against it.
+        Optional<MethodInvocationTree> maybeCheckLevel = tree.getCondition().accept(ConditionVisitor.INSTANCE, state);
+        if (!maybeCheckLevel.isPresent()) {
+            return Description.NO_MATCH;
+        }
+        MethodInvocationTree levelCheckInvocation = maybeCheckLevel.get();
+        LogLevel checkLevel = levelCheckLogLevel(levelCheckInvocation, state);
+        LogLevel mostSevere = tree.getThenStatement().accept(MostSevereLogStatementScanner.INSTANCE, state);
+        if (mostSevere == null) {
+            // Unable to find logging in this tree. This call likely delegates to something else which logs,
+            // but we cannot detect it.
+            return Description.NO_MATCH;
+        }
+        if (mostSevere == checkLevel) {
+            // The check matches the most severe log statement level. Keep up the great work!
+            return Description.NO_MATCH;
+        }
+        return buildDescription(tree)
+                .addFix(SuggestedFixes.renameMethodInvocation(
+                        levelCheckInvocation, mostSevere.levelCheckMethodName(), state))
+                .build();
+    }
+
+    @SuppressWarnings("PreferSafeLoggableExceptions")
+    private static LogLevel levelCheckLogLevel(MethodInvocationTree tree, VisitorState state) {
+        for (LogLevel level : LogLevel.values()) {
+            if (level.matchesLevelCheck(tree, state)) {
```

Comment:
😮A ckozak snippet where we allocate an iterator and loop instead of pre-building a composed function! _(just teasing)_